# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/events/Event.h
# Commit: 0ea945553f63
# Full Hash: 0ea945553f63659b99fd37265c5e7a764f70504d
# Author: Masayuki Nakano <masayuki@d-toybox.com>
# Date: 2024-12-03 16:21:08
# Regressor Bug: 1680669
# File Overlap Count: 1
# Description:
#   Bug 1680669 - part 6: Make methods of `MouseEvent` and static methods of `Event` aware of fractional coordinates r=smaug,emilio,webdriver-reviewers
#   
#   `screenX`, `screenY`, `clientX`, `clientY`, `x`, `y`, `offsetX` and `offsetY`
#   are now `double`.  This patch makes the event classes aware of DOM events.
#   
# ==============================================================================

diff -r 98d5a05c6c6e -r 0ea945553f63 dom/events/Event.h
--- a/dom/events/Event.h	Tue Dec 03 07:10:01 2024 +0000
+++ b/dom/events/Event.h	Tue Dec 03 07:10:05 2024 +0000
@@ -174,22 +174,82 @@
   bool Init(EventTarget* aGlobal);
 
   static const char16_t* GetEventName(EventMessage aEventType);
-  static CSSIntPoint GetClientCoords(nsPresContext* aPresContext,
-                                     WidgetEvent* aEvent,
-                                     LayoutDeviceIntPoint aPoint,
-                                     CSSIntPoint aDefaultPoint);
-  static CSSIntPoint GetPageCoords(nsPresContext* aPresContext,
-                                   WidgetEvent* aEvent,
-                                   LayoutDeviceIntPoint aPoint,
-                                   CSSIntPoint aDefaultPoint);
-  static Maybe<CSSIntPoint> GetScreenCoords(nsPresContext* aPresContext,
-                                            WidgetEvent* aEvent,
-                                            LayoutDeviceIntPoint aPoint);
-  MOZ_CAN_RUN_SCRIPT_BOUNDARY
-  static CSSIntPoint GetOffsetCoords(nsPresContext* aPresContext,
-                                     WidgetEvent* aEvent,
-                                     LayoutDeviceIntPoint aPoint,
-                                     CSSIntPoint aDefaultPoint);
+
+  /**
+   * Return clientX and clientY values for aEvent fired at aWidgetRelativePoint.
+   * If you do not want fractional values as the result, you should floor
+   * aWidgetRelativePoint and aDefaultClientPoint before calling this method
+   * like defined by the Pointer Events spec.
+   * https://w3c.github.io/pointerevents/#event-coordinates
+   * And finally round the result to the integer.
+   * Note that if you want fractional values and the source of
+   * aWidgetRelativePoint and aDefaultClientPoint is an untrusted event, the
+   * result may not be representable with floats, i.e., CSSPoint.  However, if
+   * it's trusted point, the result is representable with floats because we use
+   * CSSPoint to convert to/from app units.
+   */
+  static CSSDoublePoint GetClientCoords(
+      nsPresContext* aPresContext, WidgetEvent* aEvent,
+      const LayoutDeviceDoublePoint& aWidgetRelativePoint,
+      const CSSDoublePoint& aDefaultClientPoint);
+
+  /**
+   * Return pageX and pageY values for aEvent fired at aWidgetRelativePoint,
+   * which are client point + scroll position of the root scrollable frame. If
+   * you do not want fractional values as the result, you should floor
+   * aWidgetRelativePoint and aDefaultClientPoint before calling this method
+   * like defined by the Pointer Events spec.
+   * https://w3c.github.io/pointerevents/#event-coordinates
+   * And finally round the result to the integer.
+   * Note that if you want fractional values and the source of
+   * aWidgetRelativePoint and aDefaultClientPoint is an untrusted event, the
+   * result may not be representable with floats, i.e., CSSPoint.  However, if
+   * it's trusted point, the result is representable with floats because we use
+   * CSSPoint to convert to/from app units.
+   */
+  static CSSDoublePoint GetPageCoords(
+      nsPresContext* aPresContext, WidgetEvent* aEvent,
+      const LayoutDeviceDoublePoint& aWidgetRelativePoint,
+      const CSSDoublePoint& aDefaultClientPoint);
+
+  /**
+   * Return screenX and screenY values for aEvent fired at aWidgetRelativePoint.
+   * If aEvent does not support exposing the ref point, this returns Nothing. If
+   * you do not want fractional values as the result, you should floor
+   * aWidgetRelativePoint and aDefaultClientPoint before calling this method
+   * like defined by the Pointer Events spec.
+   * https://w3c.github.io/pointerevents/#event-coordinates
+   * And finally round the result to the integer.
+   * Note that if you want fractional values and the source of
+   * aWidgetRelativePoint and aDefaultClientPoint is an untrusted event, the
+   * result may not be representable with floats, i.e., CSSPoint.  However, if
+   * it's trusted point, the result is representable with floats because we use
+   * CSSPoint to convert to/from app units.
+   */
+  static Maybe<CSSDoublePoint> GetScreenCoords(
+      nsPresContext* aPresContext, WidgetEvent* aEvent,
+      const LayoutDeviceDoublePoint& aWidgetRelativePoint);
+
+  /**
+   * Return offsetX and offsetY values for aEvent fired at aWidgetRelativePoint,
+   * which are offset in the target element. If you do not want fractional
+   * values as the result, you should floor aWidgetRelativePoint and
+   * aDefaultClientPoint before calling this method like defined by the Pointer
+   * Events spec. https://w3c.github.io/pointerevents/#event-coordinates
+   * And finally round the result to the integer. Note that if you want
+   * fractional values and the source of aWidgetRelativePoint and
+   * aDefaultClientPoint is an untrusted event, the result may not be
+   * representable with floats, i.e., CSSPoint. However, if it's trusted point,
+   * the result is representable with floats because we use CSSPoint to convert
+   * to/from app units.
+   *
+   * Be aware, this may flush the layout.
+   */
+  MOZ_CAN_RUN_SCRIPT
+  static CSSDoublePoint GetOffsetCoords(
+      nsPresContext* aPresContext, WidgetEvent* aEvent,
+      const LayoutDeviceDoublePoint& aWidgetRelativePoint,
+      const CSSDoublePoint& aDefaultClientPoint);
 
   static already_AddRefed<Event> Constructor(EventTarget* aEventTarget,
                                              const nsAString& aType,
@@ -341,6 +401,10 @@
   already_AddRefed<EventTarget> EnsureWebAccessibleRelatedTarget(
       EventTarget* aRelatedTarget);
 
+  [[nodiscard]] MOZ_CAN_RUN_SCRIPT static nsIFrame*
+  GetPrimaryFrameOfEventTarget(const nsPresContext& aPresContext,
+                               const WidgetEvent& aEvent);
+
   mozilla::WidgetEvent* mEvent;
   RefPtr<nsPresContext> mPresContext;
   nsCOMPtr<EventTarget> mExplicitOriginalTarget;